Objects/weakrefobject.c

```diff
@@ -874,6 +875,31 @@ PyWeakref_GetObject(PyObject *ref)
 static void
 handle_callback(PyWeakReference *ref, PyObject *callback)
 {
+    /* If the garbage collector support is not properly implemented on
+     * some classes that are involved in a reference cycle, a weak
+     * reference may try to invoke a callback object that is being
+     * cleaned (tp_clear) by the garbage collector and it may be in an
+     * inconsistent state. As the garbage collector explicitly does
+     * not invoke callbacks that are part of the same cycle isolate as
+     * the weak reference (pretending that the weak reference was
+     * destroyed first), we should act in the same way here.
+     *
+     * When running the garbage collector pass over a generation, is
+     * possible to end in this function if the tp_clear of a function
+     * decrements the references of some internal object that is
+     * weak-referenced, invoking the weak reference callback that will
+     * try to call the function, which is in an incosistent state as
+     * is in the middle of its tp_clear and some internal fields may
+     * be NULL. */
+
+    if (PyObject_IS_GC(callback) && _PyObject_GC_IS_COLLECTING(callback)) {
+        PyErr_WarnEx(PyExc_RuntimeWarning, "A weak reference"
+                " was triying to execute a callback to a function that is being cleared by"
+                " the garbage collector.\n Some C extension class in the dependence"
+                " chain is probably not implementing correctly the garbage collector support.", 1);
+        return;
```

Comment:
What is the actual effect of just returning here? The callback isn't being called, but it wouldn't have been called had all the objects in the cycle had done the right thing, right? So what's the actual value of the warning?

Reply:
> So what's the actual value of the warning?
The reason we return is to not crash and then we warn that the interpreter has avoided a crash due to a bad implementation but we need to issue the warning so this does not pass silently.

Would you prefer to just crash?

Reply:
No, I'm trying to find out if there are _other_ reasons that people should fix their code, besides the specific crash that you're now turning into a warning. If the _only_ reason to fix the code is to avoid this warning, the warning just becomes noise -- especially because very often the warning isn't seen by the developer responsible for the incorrect C code, but some helpless user.

Reply:
Oh, I understand. Well, after this fix there is no hard crash, but an incorrect implementation of the garbage collector can have undefined consequences. These range from leaks because tp_traverse is not correctly implemented or some other crashes elsewhere.

What do you think we should do here? Maybe a more comprehensive message?

Reply:
I'm thinking *less* of a message, really :) The message is useful for people who have control over the C code that does the wrong thing, but to end users of the module. Maybe the exception should be a DeprecationWarning, so it's not shown by default? Perhaps we should discuss this on python-dev or discourse.